// react-class-container/lib/test/container/Container.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var ava_1 = require("ava");
var enzyme = require("enzyme");
var EnzymeAdapter = require("enzyme-adapter-react-16");
var src_1 = require("../../src");
enzyme.configure({ adapter: new EnzymeAdapter() });
ava_1.default("Should pass it's props and state to wrapped component", function (t) {
    var containerProps = {
        foo: "FOO",
        baz: 123
    };
    var containerState = {
        foo: "foo",
        bar: "BAR"
    };
    var childProps = undefined;
    function MyComponent(props) {
        childProps = props;
        return null;
    }
    var MyContainer = /** @class */ (function (_super) {
        __extends(MyContainer, _super);
        function MyContainer() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.state = containerState;
            return _this;
        }
        return MyContainer;
    }(src_1.Container(MyComponent)));
    enzyme.render(React.createElement(MyContainer, __assign({}, containerProps)));
    t.deepEqual(childProps, __assign(__assign({}, containerProps), containerState));
});
ava_1.default("Should pass `getChildProps` to wrapped component", function (t) {
    var props = {
        foo: "FOO",
        bar: 123,
        baz: /baz/
    };
    var childProps = undefined;
    function MyComponent(props) {
        childProps = props;
        return null;
    }
    var MyContainer = /** @class */ (function (_super) {
        __extends(MyContainer, _super);
        function MyContainer() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        MyContainer.prototype.getChildProps = function () {
            return props;
        };
        return MyContainer;
    }(src_1.Container(MyComponent)));
    enzyme.render(React.createElement(MyContainer, null));
    t.deepEqual(childProps, props);
});
ava_1.default("Should pass props and state to `getChildProps`", function (t) {
    var containerProps = {
        foo: "FOO",
        baz: 123
    };
    var containerState = {
        foo: "foo",
        bar: "BAR"
    };
    var passedProps = undefined;
    var passedState = undefined;
    var MyContainer = /** @class */ (function (_super) {
        __extends(MyContainer, _super);
        function MyContainer() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.state = containerState;
            return _this;
        }
        MyContainer.prototype.getChildProps = function (props, state) {
            passedProps = props;
            passedState = state;
            return {};
        };
        return MyContainer;
    }(src_1.Container(function () { return null; })));
    enzyme.render(React.createElement(MyContainer, __assign({}, containerProps)));
    t.deepEqual(passedProps, containerProps);
    t.deepEqual(passedState, containerState);
});
ava_1.default("Should return `getChildProps(this.props, this.state)` from childProps", function (t) {
    var containerProps = {
        foo: "FOO",
        baz: 123
    };
    var containerState = {
        foo: "foo",
        bar: "BAR"
    };
    var MyContainer = /** @class */ (function (_super) {
        __extends(MyContainer, _super);
        function MyContainer() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.state = containerState;
            return _this;
        }
        return MyContainer;
    }(src_1.Container(function () { return null; })));
    var instance = enzyme.shallow(React.createElement(MyContainer, __assign({}, containerProps))).instance();
    t.deepEqual(instance.childProps, __assign(__assign({}, containerProps), containerState));
});
